mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraIndexTableHandlerTest.java

```diff
@@ -483,4 +489,62 @@ public void updateIndexOnDeleteShouldUpdateFirstUnseenWhenUnseen() throws Except
         Optional<MessageUid> actual = firstUnseenDAO.retrieveFirstUnread(MAILBOX_ID).join();
         assertThat(actual.isPresent()).isFalse();
     }
+
+    @Test
+    public void updateIndexOnAddShouldUpdateApplicableFlag() throws Exception {
+        Flags answeredFlag = new Flags(Flag.ANSWERED);
+        MailboxMessage message = mock(MailboxMessage.class);
+        when(message.createFlags()).thenReturn(answeredFlag);
+        when(message.getUid()).thenReturn(MESSAGE_UID);
+        testee.updateIndexOnAdd(message, MAILBOX_ID).join();
+
+        Flags applicableFlag = applicableFlagDAO.retrieveApplicableFlag(MAILBOX_ID).join().get();
+
+        assertThat(applicableFlag).isEqualTo(answeredFlag);
+    }
+
+    @Test
+    public void updateIndexOnFlagsUpdateShouldUnionApplicableFlag() throws Exception {
+        Flags answeredFlag = new Flags(Flag.ANSWERED);
+        MailboxMessage message = mock(MailboxMessage.class);
+        when(message.createFlags()).thenReturn(answeredFlag);
+        when(message.getUid()).thenReturn(MESSAGE_UID);
+        testee.updateIndexOnAdd(message, MAILBOX_ID).join();
+
+        testee.updateIndexOnFlagsUpdate(MAILBOX_ID, UpdatedFlags.builder()
+            .uid(MESSAGE_UID)
+            .newFlags(new Flags(Flag.DELETED))
+            .oldFlags(answeredFlag)
+            .modSeq(MODSEQ)
+            .build()).join();
+
+        Flags applicableFlag = applicableFlagDAO.retrieveApplicableFlag(MAILBOX_ID).join().get();
+
+        assertThat(applicableFlag).isEqualTo(new FlagsBuilder().add(Flag.ANSWERED, Flag.DELETED).build());
+    }
+
+    @Test
+    public void applicableFlagShouldKeepAllFlagsEvenTheMessageRemovesFlag() throws Exception {
+        Flags messageFlags = new Flags(Flag.ANSWERED);
+        messageFlags.add(Flag.DELETED);
+        messageFlags.add(Flag.DRAFT);
+
+        MailboxMessage message = mock(MailboxMessage.class);
+        when(message.createFlags()).thenReturn(messageFlags);
+        when(message.getUid()).thenReturn(MESSAGE_UID);
+
+        testee.updateIndexOnAdd(message, MAILBOX_ID).join();
+
+        messageFlags.remove(Flag.ANSWERED);
+
+        testee.updateIndexOnFlagsUpdate(MAILBOX_ID, UpdatedFlags.builder()
+            .uid(MESSAGE_UID)
+            .newFlags(messageFlags)
```

Comment:
Should be the over way arround. You are addind the flag ANSWERED here

Should be : 

```
.oldFlags(messageFlags)
.newFlags(new Flags())
```